planner/core/logical_plan_builder.go

```diff
@@ -2828,44 +2827,32 @@ func (b *PlanBuilder) buildProjectionForWindow(p LogicalPlan, expr *ast.WindowFu
 		}
 	}
 
-	lenPartition := 0
+	var partitionItems, orderItems []*ast.ByItem
 	if spec.PartitionBy != nil {
-		items = append(items, spec.PartitionBy.Items...)
-		lenPartition = len(spec.PartitionBy.Items)
+		partitionItems = spec.PartitionBy.Items
 	}
 	if spec.OrderBy != nil {
-		items = append(items, spec.OrderBy.Items...)
+		orderItems = spec.OrderBy.Items
 	}
-	projLen := len(p.Schema().Columns) + len(items) + len(expr.Args)
+
+	projLen := len(p.Schema().Columns) + len(partitionItems) + len(orderItems) + len(expr.Args)
 	proj := LogicalProjection{Exprs: make([]expression.Expression, 0, projLen)}.Init(b.ctx)
-	schema := expression.NewSchema(make([]*expression.Column, 0, projLen)...)
+	proj.SetSchema(expression.NewSchema(make([]*expression.Column, 0, projLen)...))
 	for _, col := range p.Schema().Columns {
 		proj.Exprs = append(proj.Exprs, col)
-		schema.Append(col)
+		proj.schema.Append(col)
 	}
 
-	transformer := &itemTransformer{}
-	propertyItems := make([]property.Item, 0, len(items))
-	for _, item := range items {
-		newExpr, _ := item.Expr.Accept(transformer)
-		item.Expr = newExpr.(ast.ExprNode)
-		it, np, err := b.rewrite(item.Expr, p, aggMap, true)
-		if err != nil {
-			return nil, nil, nil, nil, err
-		}
-		p = np
-		if col, ok := it.(*expression.Column); ok {
-			propertyItems = append(propertyItems, property.Item{Col: col, Desc: item.Desc})
-			continue
-		}
-		proj.Exprs = append(proj.Exprs, it)
-		col := &expression.Column{
-			ColName:  model.NewCIStr(fmt.Sprintf("%d_proj_window_%d", p.ID(), schema.Len())),
-			UniqueID: b.ctx.GetSessionVars().AllocPlanColumnID(),
-			RetType:  it.GetType(),
-		}
-		schema.Append(col)
-		propertyItems = append(propertyItems, property.Item{Col: col, Desc: item.Desc})
+	propertyItems := make([]property.Item, 0, len(partitionItems)+len(orderItems))
+	var err error
+	p, propertyItems, err = b.buildByItemsForWindow(p, proj, partitionItems, propertyItems, aggMap)
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+	lenPartition := len(propertyItems)
+	p, propertyItems, err = b.buildByItemsForWindow(p, proj, orderItems, propertyItems, aggMap)
```

Comment:
Any specific reason why we call `buildByItemsForWindow` separately for `orderItems` and `partitionItems`?

Reply:
The only reason is to get `lenPartition` since `buildByItemsForWindow` will erase some item. If we call it once for both `orderItems` and `partitionItems`. We still need to find a way to know where is the end position of `partitionItems`.